### PR TITLE
Read tab state from tab state

### PR DESCRIPTION
**Pull request by fejta**

### CODE REVIEW

**Comment by chases2** [comment]
The API still pulls state from `/prefix/testgroup` https://github.com/GoogleCloudPlatform/testgrid/blob/9a8929edd554c2b9556bfc30d602c0a073a7fac5/pkg/api/v1/state.go#L60

Tab states are at `/prefix/dashboard/tab` instead

**Comment by fejta** [comment]
Well that a) makes sense but b) is a giant bummer as it makes this significantly more complicated

**Comment by fejta** [comment]
Thanks for pointing out that I need to use the dash/tab in the URL!

This is not my most beautiful PR but it mostly gets the job done...

File: pkg/summarizer/summary_test.go [hunk 1]
```diff
@@ -334,6 +339,91 @@ func TestUpdateDashboard(t *testing.T) {
...
+			name: "interesting tab",
+			gp: groupPather{
+				configPath: *mustPath("gs://prefix/config"),
+				gridPrefix: "ignore",
+				tabPrefix:  "tabs",
+			},
+			dash:      "hi",
+			tab:       "My favorite tab!",
+			testgroup: "bad",
+			want: func() *gcs.Path {
```
Comment (chases2) [inline pkg/summarizer/summary_test.go:396]: This function is a lot of complexity to make the internals of a url object compare right. I think it would be better to just compare path.String(); the behavior is what matters

File: cluster/prod/summarizer.yaml [hunk 2]
```diff
@@ -52,3 +52,47 @@ spec:
...
+        image: gcr.io/k8s-testgrid/summarizer:v20220225-v0.0.122-3-gccb5325
+        ports:
+        - name: metrics
+          containerPort: 2112
+        args:
+        - --config=gs://k8s-testgrid/config
+        - --grid-path=tabs
+        - --json-logs
+        - --pubsub=k8s-testgrid/tab-updates
+        - --persist-queue=gs://k8s-testgrid/queue/summarizer.json
```
[the remark this answers is not in the document]
Comment (fejta) [reply cluster/prod/summarizer.yaml]: Done

File: cluster/prod/summarizer.yaml [hunk 3]
```diff
@@ -52,3 +52,47 @@ spec:
...
+        - --summary-path=summary
```
[the remark this answers is not in the document]
Comment (fejta) [reply cluster/prod/summarizer.yaml:88]: Done

**Review by chases2 — APPROVED** [review]
There are several atomic changes here:
- Teach the API to read from tab state
- Teach the summarizer to read from tab state
- Spin up a new summarizer that reads from tab state (experimentally)

In the future, it would be easier to review as a few smaller changes.

/hold

File: cluster/prod/summarizer.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (michelle192837) [reply cluster/prod/summarizer.yaml:88]: Not sure the change went rhough; it looks like both the test group and tab summarizers are writing to <bucket>/summary (and iiuc the names of the summary files aren't changing, so they will conflict)

[the remark this answers is not in the document]
Comment (fejta) [reply cluster/prod/summarizer.yaml:88]: Note that there's no `--confirm` flag right now. I wasn't planning to write to a different location, just validate that things look good in confirm mode, and then validate things in canary before prod.

File: pkg/api/v1/state.go [hunk 4]
```diff
@@ -30,56 +30,62 @@ import (
...
 	statepb "github.com/GoogleCloudPlatform/testgrid/pb/state"
+	"github.com/GoogleCloudPlatform/testgrid/pkg/tabulator"
 	"github.com/GoogleCloudPlatform/testgrid/util/gcs"
 	"github.com/golang/protobuf/ptypes/timestamp"
 	"github.com/gorilla/mux"
 )
 
 // findDashboardTab locates dashboard tab in config
-func findDashboardTab(cfg *configpb.Configuration, dashboardKey string, tabKey string) (*configpb.DashboardTab, error) {
+func findDashboardTab(cfg *configpb.Configuration, dashboardKey string, tabKey string) (string, *configpb.DashboardTab, error) {
```
[the remark this answers is not in the document]
Comment (fejta) [reply pkg/api/v1/state.go:41]: Done

File: pkg/api/v1/state_test.go [hunk 5]
```diff
@@ -145,6 +151,79 @@ func TestDecodeRLE(t *testing.T) {
...
+			},
+			endpoint:         "dashboard1/tabs/tab1/rows",
+			expectedResponse: `{}`,
+			expectedCode:     http.StatusOK,
+		},
+	}
+	RunTestsAgainstEndpoint(t, "/dashboards/", tests)
+}
+
+// TODO(fejta): test request/response
```
[the remark this answers is not in the document]
Comment (fejta) [reply pkg/api/v1/state_test.go:226]: Not sure what that is supposed to gain me...

TODOs are more reminders, less recipes. If a person cannot infer what to do, they are probably ill equipped to address the TODO.

File: pkg/summarizer/summary_test.go (hunk 1 above)
Comment (fejta) [reply pkg/summarizer/summary_test.go:396]: Added a TODO to implement a https://pkg.go.dev/github.com/google/go-cmp/cmp#Transform, which a good way to address this concern  (this is what protocmp.Transform() implements I suspect).

My main goal here is to fail something here when we change the tabulator function that this relies on. It is only one test so it doesn't seem necessary to change the rest of them when I can just adapt this one.

File: pkg/api/v1/state_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (chases2) [reply pkg/api/v1/state_test.go:226]: This reminder feels unkind to your future self. "Remove HTTP server" makes it so much clearer what you were thinking without looking up the PR it came from.

I could not infer what to do from this without the context of the PR it came from or our previous conversations.

File: pkg/summarizer/summary_test.go (hunk 1 above)
Comment (chases2) [reply pkg/summarizer/summary_test.go:396]: I'm not sure I understand. Why would you want to avoid changing the other tests in `TestGroupPatherLookup`? Making "want" a `string` instead of `*gcs.Path` seems clear, unless you're worried about some side effect dependent on the internals of `gcs.Path`. And if you are, those tests should probably go with that package.

**Review by chases2 — APPROVED** [review]

**Comment by fejta** [comment]
> I'm not sure I understand. Why would you want to avoid changing the other tests in TestGroupPatherLookup? Making "want" a string instead of *gcs.Path seems clear, unless you're worried about some side effect dependent on the internals of gcs.Path. And if you are, those tests should probably go with that package.


Just for efficiency. It is not hard to fudge the one case to work correctly, and then I don't have to worry about any side effects.

OTOH, it takes time to change all the existing cases to use a string, and then update the test to handle the string, and decide whether and how to deal with edge cases like a `nil` `*gcs.Path`... Should I just make this a "" or does that risk becoming equivalent with `gcs.Path{}` etc.

At a high level I agree with you, and I think the correct place to try and encapsulate all that thinking in a single place like the cmp.Transform. Then I can reuse it in elsewhere in the future.
